AmbireTech/ambire-common
# Enhancement/ Improve portfolio error banner UX

## Changes:
- Add: `selectedAccount` banner tests- the logic is quite complicated and we didn't have any test cases which is why I added some
- Change: Display a portfolio error banner only if there isn't a successful update in the last 10 minutes or if the user force updates the portfolio
- Change: Display defi position error banners only if there isn't a successful update in the last 10 minutes and ONLY for networks and providers the user has positions on (force update isn't a trigger here)
- Update: Portfolio lib error structure to contain `level` as some errors are critical to the response (velcro error), some are important (portfolio price error) and some should only be stored for DX (individual token errors)

## How to test:
1. In `providers` ctrl, row 56:
    ```
    this.providers[networkId].isWorking = true
    this.emitUpdate()

    return
    ```
2. In `background.ts` (the extension), row 725, add `mainCtrl.defiPositions.updatePositions()`
3. In `banners.ts`, row 18, update `const TEN_MINUTES = 10 * 60 * 1000` to `const TEN_MINUTES = 0.5 * 60 * 1000`

### What do these changes affect?
1. This allows us to block RPCs from the network tab in devtools without flagging RPCs as `isWorking=false` because when they are flagged as such portfolio and defi banners are replaced with an RPC banner
2. The only way to manually update defi positions is to use the refresh button on the dashboard, but that triggers a force update of the portfolio, which causes a portfolio banner to be displayed, even if there is fresh data. To manually update defi positions and the portfolio, we add this code, which is executed when the extension popup is opened.
3. Portfolio and defi positions data is "valid" for 10 minutes, which makes it hard to test. Reduce it to the time you want (30sec in this example)

### Example test steps:
1. Load an account with defi positions
2. Block the RPC of a network, where the account has defi positions
3. Reload the extension and open it- Expect a portfolio and defi position banner(s)
4. Unblock the RPC and open the extension (don't reload it)- expect no banners
5. Block an RPC on any network
6. Load an account (2) you haven't loaded- expect all banners, even if the account doesn't have defi positions as there is no data for the user in `networksWithPositions`
7. Unblock the RPC, load the same account (2), block the RPC again, reload the extension and expect only relative defi positions banners
... try other steps

### Important:
Opening the extension popup doesn't always reload defi positions and the portfolio. Check `#getCanSkipUpdate` for more info.
